README
NAME

    Dancer2::Session::PSGI - Dancer2 session storage via
    Plack::Middleware::Session

VERSION

    version 0.008

SYNOPSIS

        use Dancer2;
    
        setting( session => 'PSGI' );
    
        get '/' => sub {
            my $count = session("counter");
            session "counter" => ++$count;
            return "This is my ${count}th dance";
        };

DESCRIPTION

    This module implements implements a session factory for Dancer2 that
    uses Plack::Middleware::Session for session management.

CONFIGURATION

    The setting session should be set to PSGI in order to use this session
    engine in a Dancer2 application.

    The default cookie name is plack_session. Refer to "Session_engine" in
    Dancer2::Config if you need to modify this.

ACKNOWLEDGEMENTS

    The methods required by Dancer2::Core::Role::SessionFactory were
    heavily based on Dancer2::Session::Cookie by David Golden.

AUTHOR

    Russell Jenkins <russellj@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Russell Jenkins.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision History for Dancer2::Session::PSGI

0.008     2014-10-16T08:26:41EST+1100
  * Require Plack::Middleware::Session v0.23 as it is the first release
    of Middleware::Session that requires a secret.

0.007     2014-10-14T14:38:14EST+1100
  * Access to PSGI env is via engine request attr, not context.
    (Restore functionality with Dancer2 v0.152000)

0.006     2014-06-23T11:07:19EST+1000
  * Updating metadata. No functional changes.

0.005     2014-06-23T10:17:37EST+1000
  * Ensure changes session expiry are passed back to the psgi env.

0.004     2014-04-18T20:19:40EST+1000
  * Update test prerequisites: Tests depend on Plack, as Plack::Test does
    not have version information.

0.003     2014-04-13T20:30:18EST+1000
  * Update test prerequisites:
    Plack::Test >= 1.0029 for simplified OO interface
    (Reported by @haarg)

0.002     2014-04-01T06:36:07EST+1100
  * Revise test prerequisites.

0.001     2014-03-31T00:26:03EST+1100
  * Initial release


lib/Dancer2/Session/PSGI.pm
package Dancer2::Session::PSGI;

# ABSTRACT: Dancer2 session storage via Plack::Middleware::Session

use Moo;
with 'Dancer2::Core::Role::SessionFactory';

our $VERSION = '0.008'; # VERSION

#-----------------------------------------#
# Alter SessionFactory attribute defaults
#-----------------------------------------#

has '+cookie_name' => ( default => 'plack_session' );

#-----------------------------------------#
# SessionFactory implementation methods
#-----------------------------------------#

# Get middleware session hash
sub _retrieve {
    shift->request->env->{'psgix.session'};
}

# Put data back/into middleware session hash
sub _flush {
    my ( $self, $id, $data ) = @_;
    $self->request->env->{'psgix.session'} = $data;
}

# Middleware handles cookie expiry
sub _destroy { return }

# Its the responsibility of Plack::Middleware::Session for
# tracking other sessions (we know nothing about them).
# So return an empty list.
sub _sessions { return [] }

#-----------------------------------------#
# Overridden methods from SessionFactory
#-----------------------------------------#

# Middleware sets the cookie.
sub set_cookie_header {
    my ( $self, %params ) = @_;

    my $session = $params{session};
    my $options = $self->request->env->{'psgix.session.options'};

    if ( my $expires = $session->expires ) {
        if ( $session->expires < time ) { # Cookie has expired.
            $options->{expire} = 1;
        }
        else { # Pass upstream the cookie expire time
            $options->{expires} = $expires;
        }
    }
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Dancer2::Session::PSGI - Dancer2 session storage via Plack::Middleware::Session

=head1 VERSION

version 0.008

=head1 SYNOPSIS

    use Dancer2;

    setting( session => 'PSGI' );

    get '/' => sub {
        my $count = session("counter");
        session "counter" => ++$count;
        return "This is my ${count}th dance";
    };

=head1 DESCRIPTION

This module implements implements a session factory for Dancer2 that uses
L<Plack::Middleware::Session> for session management.

=for Pod::Coverage method_names_here
set_cookie_header

=head1 CONFIGURATION

The setting B<session> should be set to C<PSGI> in order to use this session
engine in a Dancer2 application.

The default cookie name is C<plack_session>. Refer to
L<Dancer2::Config/Session_engine> if you need to modify this.

=head1 ACKNOWLEDGEMENTS

The methods required by L<Dancer2::Core::Role::SessionFactory> were
heavily based on L<Dancer2::Session::Cookie> by David Golden.

=head1 AUTHOR

Russell Jenkins <russellj@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Russell Jenkins.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Dancer2 session storage via Plack::Middleware::Session",
  "AUTHOR" => "Russell Jenkins <russellj\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dancer2-Session-PSGI",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Dancer2::Session::PSGI",
  "PREREQ_PM" => {
    "Dancer2" => "0.152000",
    "Dancer2::Core::Role::SessionFactory" => 0,
    "Moo" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer2" => "0.152000",
    "File::Spec" => 0,
    "HTTP::Cookies" => 0,
    "HTTP::Request::Common" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Plack" => "1.0029",
    "Plack::Builder" => 0,
    "Plack::Middleware::Session" => "0.23",
    "Plack::Test" => 0,
    "Test::More" => "0.96",
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.008",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dancer2" => "0.152000",
  "Dancer2::Core::Role::SessionFactory" => 0,
  "File::Spec" => 0,
  "HTTP::Cookies" => 0,
  "HTTP::Request::Common" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moo" => 0,
  "Plack" => "1.0029",
  "Plack::Builder" => 0,
  "Plack::Middleware::Session" => "0.23",
  "Plack::Test" => 0,
  "Test::More" => "0.96",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



